lsp-mode.el

```diff
@@ -4999,27 +5086,49 @@ Others: TRIGGER-CHARS"
                        'lsp-completion-prefix prefix)
                (text-properties-at 0 candidate))
               ((&CompletionItem :label :insert-text? :text-edit? :insert-text-format? :additional-text-edits?)
-               item))
+               item)
+              main-edit-region
+              (change-tracker (lambda (changes)
+                                (when-let ((buf-changes (plist-get changes (current-buffer)))
+                                           (change-starts (nconc (mapcar #'cl-first buf-changes)
+                                                                 (cl-first main-edit-region)))
+                                           (change-start (apply #'min change-starts))
+                                           (change-ends (nconc (mapcar #'cl-second buf-changes)
+                                                               (cl-second main-edit-region)))
+                                           (change-end (apply #'max change-ends)))
+                                  (setq main-edit-region `(,change-start ,change-end))))))
         (cond
          (text-edit?
+          (setq main-edit-region (lsp--range-to-region (lsp:text-edit-range text-edit?)))
           (apply #'delete-region markers)
           (insert prefix)
           (lsp--apply-text-edit text-edit?))
          ((or insert-text? label)
+          (setq main-edit-region (cons start-point (point)))
           (apply #'delete-region markers)
           (insert prefix)
           (delete-region start-point (point))
           (insert (or insert-text? label))))
 
-        (when (eq insert-text-format? 2)
+        (when (equal insert-text-format? lsp/insert-text-format-snippet)
           (let ((yas-indent-line (lsp--indent-snippets?)))
             (yas-expand-snippet
              (lsp--to-yasnippet-snippet (buffer-substring start-point (point)))
              start-point
              (point))))
 
-        (when (and lsp-completion-enable-additional-text-edit additional-text-edits?)
-          (lsp--apply-text-edits additional-text-edits?))
+        (when lsp-completion-enable-additional-text-edit
+          (if (or (get-text-property 0 'lsp-completion-resolved candidate)
+                  additional-text-edits?)
+              (lsp--apply-additional-text-edits additional-text-edits? main-edit-region)
+            (add-hook 'lsp-on-change-functions change-tracker)
+            (lsp--resolve-completion-async
+             item
+             (lambda (resolved-item)
+               (remove-hook 'lsp-on-change-functions change-tracker)
```

Comment:
can we use directly before-change-functions? Seems like it should be buffer-local subscription. Can you generalize it so we can use it for on type formatting as well?

Reply:
We might be able to use `before-change-functions` directly, but it's not debounced and not obeying to `lsp-inhibit-hook`. I don't think we need to apply additional text edits immediately, it already applied asynchronously, so wait a little bit more to a debounce check-point would also be fine.
The new hook `lsp-on-change-functions` can be used in type formatting too, we will need to make change tracker a separate function, I will do that in another PR.

Reply:
We don't need to debounce since the only thing that function does is to store information need for checking if a particular set of edits is invalidated, no?